File: Matlab/master.m
clear all
init;

%synindex = [4960 4970 4980 4990 5000 9960 9970 9980 9990 10000];
synindex = [10 15 20 25];
synData = cell(length(synindex),1);
for i = 1:mc.nrun  
    
    tic;   
     %% update zHH
     [z_HH,z_HH_Individuals] = samplezHHwithHHnewv1_2HHvar(para.phi,orig.dataT,...
         para.w,para.pi,orig.SS,para.HHdata_all(:,1:orig.n)',para.lambda{1},para.lambda{2}, rand(orig.n,1));

     %% update zIndividual
     z_Individuals = samplezmemberv1(para.phi,orig.dataT,para.w,z_HH,orig.HHserial,rand(size(orig.dataT,2),1));
    

        %% generate impossible house hold
        [z_Individual_extra,z_HH_extra,IndividualData_extra,HHdata_extra,para.hh_size_new,current_synData] = ...
        GetImpossibleHouseholds(orig.d,orig.ACS_count,para.lambda,para.w,...
        para.phi,para.pi,hyper.howmany,orig.n, ismember(i,synindex));
        if ismember(i,synindex) 
            synData{find(synindex ==i)} = current_synData(1:8,:);
        end
        %% combine data and indicators
        para.z_HH_all = [z_HH z_HH_extra];
        para.HHdata_all = orig.HHdataorigT; para.HHdata_all(2,:) = para.HHdata_all(2,:) - 1;
        para.HHdata_all = [para.HHdata_all HHdata_extra];
        para.IndividualData_all = [orig.origdata(:,1:8)' IndividualData_extra];
        
        %row 1 for K groups and row 2 for L groups
        para.z_Individual_all = [[z_HH_Individuals;z_Individuals] z_Individual_extra];
       
        clear z_HH z_HH_extra z_Individuals z_HH_Individuals z_Individual_extra IndividualData_extra HHdata_extra

        %% update phi
        para.phi = UpdatePhi(hyper.K,hyper.L,orig.p,orig.d,orig.maxd,...
                    para.IndividualData_all,para.z_Individual_all);
        %% update w
        [para.w,para.v] = UpdateW(para.beta,para.z_Individual_all, hyper.K, hyper.L);
  
    
     %% update lambda
    [para.lambda] = UpdateLambda(hyper.dHH,hyper.K,para.z_HH_all,para.HHdata_all);
    
     %% update pi
    [para.pi,para.u] = UpdatePi(para.alpha,para.z_HH_all,hyper.K);
        
    
    
     %% update alpha
    para.alpha = UpdateAlpha(hyper.aa,hyper.ab,para.u);
    
    %% update beta
    para.beta = UpdateBeta(hyper.ba,hyper.bb,para.v);
    
    postsave;
    
%    if mod(i,1000) == 0
%        save -v7.3 last;
%     end
end
   

File: R/SynthHousehold/master.R
library(SynthHousehold)

#set up data set
data(household)
orig <- initData(household)

#mcmc parameters
mc <- list(nrun = 25, burn = 20, thin = 1)
mc$eff.sam <- (mc$nrun-mc$burn)/mc$thin

#hyper parameters
#aa ab: gamma hyperparameters for alpha
hyper <- list(K=40 , L=15, aa=0.25, ab=0.25, ba=0.25,bb=0.25, dHH = c(2,3), blocksize = 10000)

para <- initParameters(orig,hyper)
output <- initOutput(orig,hyper,mc)

synindex <- c(9910,9920,9930,9940,9950,9960,9970,9980,9990,10000)
synData <- list()

for (i in 1:mc$nrun) {
  # update zHH
  z_household <- samplezHH(para$phi,orig$dataT,para$w,para$pi,orig$SS,t(para$HHdata_all[,1:orig$n]),
            para$lambda[[1]],para$lambda[[2]])

  # update zIndividual
  z_Individuals <- samplezmember(para$phi,orig$dataT,para$w,z_household$z_HH,orig$HHserial)
  data.extra <- GetImpossibleHouseholds(orig$d,orig$ACS_count,para$lambda,para$w,para$phi,
                  para$pi,hyper$blocksize,orig$n,is.element(i,synindex))
  if (is.element(i,synindex)) {
    synData[[which(synindex ==i)]] <- data.extra$synIndividuals_all[1:8,]
  }
    #combine data and indicators
    para$z_HH_all <- c(z_household$z_HH, data.extra$z_HH_extra)
    para$HHdata_all <- orig$HHdataorigT
    para$HHdata_all[2,] <- para$HHdata_all[2,] -1
    para$HHdata_all <- rbind(z_household$z_HH_Individuals,z_Individuals)
    para$IndividualData_all <- cbind(t(orig$origdata[,1:8]),data.extra$IndividualData_extra)

    #row 1 for K groups and row 2 for L groups
    temp <- rbind(z_household$z_HH_Individuals,z_Individuals)
    para$z_Individual_all  <- cbind(temp,data.extra$HHdata_individual_extra)

    # update phi
    para$phi <- UpdatePhi(para$IndividualData_all,para$z_Individual_all, hyper$K,hyper$L,orig$p,orig$d,orig$maxd);

    W <- UpdateW(para$beta,para$z_Individual_all, hyper$K, hyper$L)
    para$w <- W$w
    para$v <- W$v


  #postsave
}
